### ISSUE TITLE
storage: disappearing files with rocksdb `MemEnv`

### ISSUE THREAD

**Issue by asubiotto** [cockroachdb/cockroach#45505]
Reproduction:
1) Create a new logic test file in `pkg/sql/logictest/testdata/logic_test/$FNAME`
2) Paste this into the logic test file:
```
# LogicTest: fakedist-vec-disk

# Regression test for #40732.
statement ok
CREATE TABLE t40732 AS SELECT g::INT8 AS _int8,
                              g::FLOAT8 AS _float8,
                              '2001-01-01'::DATE
                              + g AS _date,
                              g % 2 = 1 AS _bool,
                              g::DECIMAL AS _decimal,
                              g::STRING AS _string,
                              g::STRING::BYTES AS _bytes
                         FROM generate_series(1, 5) AS g

statement ok
INSERT INTO t40732 DEFAULT VALUES

query I
SELECT *
  FROM (
          SELECT tab_1541._int8 AS col_2976
            FROM t40732 AS tab_1538
            JOIN t40732 AS tab_1539
            JOIN t40732 AS tab_1540 ON
                  tab_1539._float8 = tab_1540._float8
            JOIN t40732 AS tab_1541 ON
                  tab_1540._int8 = tab_1541._int8 ON
                  tab_1538._float8 = tab_1540._float8,
                 t40732 AS tab_1542
           WHERE tab_1542._bool > tab_1540._bool
       )
ORDER BY col_2976
----
2
2
2
4
4
4
```
3) Run `make stress PKG=./pkg/sql/logictest TESTS="TestLogic/fakedist-vec-disk/$FNAME`

The test errors out with `file does not exist`. The access pattern of the file is `CreateFile`, `Write` x n, then an `OpenFile` which fails. Running this test under stress with pebble as the storage engine runs fine so the problem seems to be in the rocksdb in-memory storage engine.

**Comment by sumeerbhola** [comment]
I must be doing something wrong:

```
$ make test PKG=./pkg/sql/logictest TESTS="TestLogic/fakedist-vec-disk/missing"
Running make with -j16
GOPATH set to /Users/sumeer/go
Cleaning old generated files.
GOFLAGS= go test   -tags ' make x86_64_apple_darwin19.3.0' -ldflags '-X github.com/cockroachdb/cockroach/pkg/build.typ=development -extldflags "" -X "github.com/cockroachdb/cockroach/pkg/build.tag=v20.1.0-beta.1-417-gc331fae662-dirty" -X "github.com/cockroachdb/cockroach/pkg/build.rev=c331fae662cde9a1f59afd0305a392775fc41b28" -X "github.com/cockroachdb/cockroach/pkg/build.cgoTargetTriple=x86_64-apple-darwin19.3.0"  ' -run "TestLogic/fakedist-vec-disk/missing"  -timeout 20m ./pkg/sql/logictest 
I200227 20:51:15.796805 1 rand.go:85  Random seed: -8916779150996257147
test logs left over in: /var/folders/rm/0t5c352977v4t7rz3qzjy1x40000gn/T/logTestLogic873502921
--- FAIL: TestLogic (0.50s)
    test_log_scope.go:77: test logs captured to: /var/folders/rm/0t5c352977v4t7rz3qzjy1x40000gn/T/logTestLogic873502921
    test_log_scope.go:58: use -show-logs to present logs inline
    --- FAIL: TestLogic/fakedist-vec-disk (0.00s)
        --- FAIL: TestLogic/fakedist-vec-disk/missing (0.44s)
            logic.go:1920: 
                
                testdata/logic_test/missing:18: SELECT *
                  FROM (
                          SELECT tab_1541._int8 AS col_2976
                            FROM t40732 AS tab_1538
                            JOIN t40732 AS tab_1539
                            JOIN t40732 AS tab_1540 ON
                                  tab_1539._float8 = tab_1540._float8
                            JOIN t40732 AS tab_1541 ON
                                  tab_1540._int8 = tab_1541._int8 ON
                                  tab_1538._float8 = tab_1540._float8,
                                 t40732 AS tab_1542
                           WHERE tab_1542._bool > tab_1540._bool
                       )
                ORDER BY col_2976
                expected success, but found
                (XX000) internal error: unexpected error from the vectorized runtime: projectingBatch doesn't support ColVecs()
                error.go:77: in func1()
                DETAIL: stack trace:
                github.com/cockroachdb/cockroach/pkg/sql/colexec/execerror/error.go:77: func1()
                runtime/panic.go:522: gopanic()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/execerror/error.go:159: VectorizedInternalPanic()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/simple_project.go:51: ColVecs()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/allocator.go:77: RetainBatch()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/external_sort.go:365: Next()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/sort.go:135: spool()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/sort.go:265: Next()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/external_sort.go:214: Next()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/disk_spiller.go:205: Next()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/disk_spiller.go:220: Next()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:45: Next()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:149: next()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:140: nextAdapter()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/execerror/error.go:91: CatchVectorizedRuntimeError()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:177: Next()
                github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:170: Run()
                github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:749: Run()
                github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:370: Run()
                github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:377: Run()
                github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:987: PlanAndRun()
                github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:795: execWithDistSQLEngine()
                github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:709: dispatchToExecutionEngine()
                github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:412: execStmtInOpenState()
                github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:93: execStmt()
                github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1355: execCmd()
                github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1284: run()
                github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:483: ServeConn()
                github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:566: func1()
                runtime/asm_amd64.s:1337: goexit()
                
                NOTE: internal errors may have more details in logs. Use -show-logs.
            logic.go:1674: 
                 pq: internal error: unexpected error from the vectorized runtime: projectingBatch doesn't support ColVecs()
FAIL
FAIL	github.com/cockroachdb/cockroach/pkg/sql/logictest	2.968s
```

**Comment by asubiotto** [comment]
Do you have the latest master checked out? https://github.com/cockroachdb/cockroach/pull/45318 fixes that error.

**Comment by sumeerbhola** [comment]
That fixes that error.

I think we've screwed up the slice/string memory somehow, which may also explain #45115 

The file is there but the std::map<std::string, MemFile*> is not in sorted order, and the size is inconsistent with what the iterator is finding. The following is a failure to find a particular filename in the map with size 14, and the filenames printed in iterator order -- the filename is there (entry 14 but that would make it the 15th entry!) but the ordering of the names is wrong.

```
0xc4da200: NewRandomAccessFile: f8cd6c0d-4ec8-4efd-b448-0cc0e3446b35/3ac9ed11-1477-46c0-9791-8466b3876e1f/0 not found 14
0: 
1: /
2: /000003.log
3: /000006.log
4: /000007.sst
5: /CURRENT
6: /IDENTITY
7: /LOCK
8: /MANIFEST-000008
9: /OPTIONS-000005
10: f8cd6c0d-4ec8-4efd-b448-0cc0e3446b35/3ac9ed11-1477-46c0-9791-8466b3876e1f/0
11: f8cd6c0d-4ec8-4efd-b448-0cc0e3446b35
12: f8cd6c0d-4ec8-4efd-b448-0cc0e3446b35/2c70857e-c562-4f89-8181-3dac75743dc9
13: f8cd6c0d-4ec8-4efd-b448-0cc0e3446b35/2c70857e-c562-4f89-8181-3dac75743dc9/0
14: f8cd6c0d-4ec8-4efd-b448-0cc0e3446b35/3ac9ed11-1477-46c0-9791-8466b3876e1f
```

**Comment by sumeerbhola** [comment]
This was simply a missing `MutexLock` in `MockEnv::CreateDir` https://github.com/facebook/rocksdb/blob/master/env/mock_env.cc#L588-L594
It should fix #45115 too.
